jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketCoreSession.java

```diff
@@ -422,7 +428,31 @@ public void demand(long n)
     {
         if (!demanding)
             throw new IllegalStateException("FrameHandler is not demanding: " + this);
-        connection.demand(n);
+
+        LongConsumer consumer = popDemandHandler();
+        if (consumer != null)
+            consumer.accept(n);
+        else
+            connection.demand(n);
+    }
+
+    public LongConsumer popDemandHandler()
+    {
+        LongConsumer consumer = null;
+        try (AutoLock l = _lock.lock())
+        {
+            if (!_demandHandlers.isEmpty())
+                consumer = _demandHandlers.pop();
+        }
+        return consumer;
+    }
+
+    public void pushDemandHandler(LongConsumer demand)
```

Comment:
I'm not opposed to a demand interception architecture.... as ultimately that is what we are doing in the jetty-12 HttpStream contract.   However, at the very least it needs a lot of javadoc to explain how to use it and why it is needed.

But I'm also very cautious that it doesn't go hand in hand with the interception of the actual content... again as it is done in the jetty-12 HttpStream contract.

Reply:
I think you need to look at adding demand interception to the Extension class, perhaps as an optional interface extension.   Then WebSocketCoreSession, which it builds its stack of extentions that intercept incoming frames, it can also build a stack of demand interceptors (handlers), that will always be there - not added and removed.    Thus the inflate extension will always intercept demand and only pass it on if it has not more bytes to inflate into a synthetic frame.

Reply:
> at the very least it needs a lot of javadoc to explain how to use it and why it is needed.

Sure I can add some javadoc. The `DemandHandler` is internal mechanism and is not exposed through the `CoreSession` interface. So the only way to have access to this is to downcast the `CoreSession` to an internal class.

> I think you need to look at adding demand interception to the Extension class, perhaps as an optional interface extension.

This is not possible, it needs to be on the WebSocketCoreSession or the WebSocketConnection, the application has no interaction with the Extensions it only calls demand on the session. I originally had a permanent `DemandListener` to intercept demand, so that is possible but it still needs to be registered through the `WebSocketCoreSession`.

Reply:
> Then WebSocketCoreSession, which it builds its stack of extentions that intercept incoming frames, it can also build a stack of demand interceptors (handlers), that will always be there - not added and removed. Thus the inflate extension will always intercept demand and only pass it on if it has not more bytes to inflate into a synthetic frame.

What's currently in this PR is a minimal, implementation-only, fix for the issue.
While perhaps a more generic (and publicly available) mechanism would be desired, I think for this release we should stick with this minimal change, as what you propose is a larger change that still feels like an afterthought (introducing new methods to `Extension` that should really be a new parameter in `IncomingFrames.onFrame()` instead, etc.).
I like better a re-design for Jetty 12 and a minimal change now.
Then if we really need to add a method to `Extension` we can discuss it more carefully for a future Jetty 10 release.

Reply:
@sbordet the current PR doesn't need to be a perfect fix and we can land on a final API in 12.  But it should not be a step in the wrong direction.   I think the approach here is just wrong: a dynamic stack of interceptors and explicit demand bypass API!        We should at least fix it in a style that is likely to be similar to what we ultimately implement - ie a fixed set of demand interceptors that is initialized once the Extentions are known, that will always be called for all demand calls.   I see no reason not to try to do it right now.